### PR TITLE
Request hash on filterstate

### PR DESCRIPTION
**Pull request by stevenzzzz**
Description: Add a new hash method based on Hashable object in FilterState. 

Risk Level: LOW (new method)
Testing: unit tests
Docs Changes: N/A 
Release Notes: N/A
[Fixes] #9701

### CODE REVIEW

File: include/envoy/http/hash_policy.h [hunk 1]
```diff
@@ -2,12 +2,19 @@
...
+#include "envoy/stream_info/filter_state.h"
 
 #include "absl/types/optional.h"
 
 namespace Envoy {
 namespace Http {
 
+class Hashable {
+public:
+  virtual absl::optional<uint64_t> hash() const { return absl::nullopt; };
```
Comment (htuch) [inline include/envoy/http/hash_policy.h]: Why isn't this `PURE`? It seems that any hashable object should be able to always produce a hash?

Comment (stevenzzzz) [reply include/envoy/http/hash_policy.h]: Removed.

File: include/envoy/http/hash_policy.h [hunk 2]
```diff
@@ -11,7 +11,7 @@ namespace Http {
 
 class Hashable {
 public:
-  virtual absl::optional<uint64_t> hash() const { return absl::nullopt; };
+  virtual absl::optional<uint64_t> hash() const PURE;
```
Comment (htuch) [inline include/envoy/http/hash_policy.h]: Does it still need to be optional as well?

Comment (stevenzzzz) [reply include/envoy/http/hash_policy.h]: I want to keep it "relaxed", in the case when there is no data available (say, header not available yet), return the same value as the evaluate() suggests.
We can definitely require subclasses of the Hashable class to return 0, but optional maybe aligns better with the existing interface?

**Pull request merged**